infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -143,102 +142,147 @@ public void checkpointState() {
+            latestGlobalOffset = Math.max(latestGlobalOffset, lastIssued == null ? Long.MIN_VALUE : lastIssued);
+        }
+        if (!hit) {
+            latestGlobalOffset = -1L; //no token ever issued
+        }
+        if (req.getStreams().size() == 0) {
+            latestGlobalOffset = globalIndex.get() - 1;
+        }
+        r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(
+                new TokenResponse(latestGlobalOffset, Collections.emptyMap(), requestLatestStreamOffsets.build())));
+    }
+
+    /**
+     * Service an incoming token request.
+     */
+    @ServerHandler(type=CorfuMsgType.TOKEN_REQ)
+    public synchronized void tokenRequest(CorfuPayloadMsg<TokenRequest> msg,
+                                          ChannelHandlerContext ctx, IServerRouter r) {
+        TokenRequest req = msg.getPayload();
+        if (req.getNumTokens() == 0) {
+            returnLatestOffsets(msg, ctx, r);
         } else {
             long thisIssue = globalIndex.getAndAdd(req.getNumTokens());
             if (req.getStreams() == null) {
                 r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(
                         new TokenResponse(thisIssue, Collections.emptyMap(), Collections.emptyMap())));
                 return;
             }
+            // If the request is a transaction resolution request, then check if it should abort.
             if (req.getTxnResolution()) {
-                // Then also need a read timestamp.
-                long timestamp = req.getReadTimestamp();
-                if (timestamp != -1L) {
-                    AtomicBoolean abort = new AtomicBoolean(false);
-                    for (UUID id : req.getStreams()) {
-                        if (abort.get())
-                            break;
-                        lastIssuedMap.compute(id, (k, v) -> {
-                            if (v == null) {
-                                return null;
-                            } else {
-                                if (v > timestamp) {
-                                    log.debug("Rejecting request due to {} > {} on stream {}", v, timestamp, id);
-                                    abort.set(true);
-                                }
-                            }
-                            return v;
-                        });
-                    }
-                    if (abort.get()) {
-                        globalIndex.getAndAdd(-req.getNumTokens());
-                        r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(
-                                new TokenResponse(-1L, Collections.emptyMap(), Collections.emptyMap())));
-                        return;
-                    }
+                if (!txnResolution(req.getReadTimestamp(), req.getReadSet())) {
+                    // If the txn aborts, then DO NOT hand out a token.
+                    globalIndex.getAndAdd(-req.getNumTokens());
```

Comment:
I see. Indeed, I woke up today about to check that gettoken is synchronized, because it is needed in order to make global and stream index allocation atomic.

But then 

(a) why is globalIndex declared AtomicLong? 

and (b) what about other methods not declared synchronized interfering? In particular, could a soft reset happening in between allocating a global-index and streams-indexes cause inconsistency?